File: music/src/test/java/io/xj/music/PitchClassTest.java
// Copyright (c) 2017, Outright Mental Inc. (http://outright.io) All Rights Reserved.
package io.xj.music;

import org.junit.Assert;
import org.junit.Test;

import static io.xj.music.AdjSymbol.Flat;
import static io.xj.music.AdjSymbol.None;
import static io.xj.music.AdjSymbol.Sharp;
import static org.junit.Assert.assertEquals;

public class PitchClassTest {


  @Test
  public void NameOfTest() throws Exception {
    assertPitchClassOf("C", PitchClass.C, "C", "C");
    assertPitchClassOf("C#", PitchClass.Cs, "C#", "Db");
    assertPitchClassOf("Cb", PitchClass.B, "B", "B");
    assertPitchClassOf("D", PitchClass.D, "D", "D");
    assertPitchClassOf("D#", PitchClass.Ds, "D#", "Eb");
    assertPitchClassOf("D♭", PitchClass.Cs, "C#", "Db");
    assertPitchClassOf("E", PitchClass.E, "E", "E");
    assertPitchClassOf("E#", PitchClass.F, "F", "F");
    assertPitchClassOf("E♭", PitchClass.Ds, "D#", "Eb");
    assertPitchClassOf("F", PitchClass.F, "F", "F");
    assertPitchClassOf("F#", PitchClass.Fs, "F#", "Gb");
    assertPitchClassOf("F♭", PitchClass.E, "E", "E");
    assertPitchClassOf("G", PitchClass.G, "G", "G");
    assertPitchClassOf("G♯", PitchClass.Gs, "G#", "Ab");
    assertPitchClassOf("Gb", PitchClass.Fs, "F#", "Gb");
    assertPitchClassOf("A", PitchClass.A, "A", "A");
    assertPitchClassOf("A#", PitchClass.As, "A#", "Bb");
    assertPitchClassOf("Ab", PitchClass.Gs, "G#", "Ab");
    assertPitchClassOf("B", PitchClass.B, "B", "B");
    assertPitchClassOf("B#", PitchClass.C, "C", "C");
    assertPitchClassOf("E♭", PitchClass.Ds, "D#", "Eb");
    assertPitchClassOf("Bb", PitchClass.As, "A#", "Bb");
    assertPitchClassOf("z", PitchClass.None, "-", "-");
    assertPitchClassOf("zzzz", PitchClass.None, "-", "-");
  }

  @Test
  public void NameStepTest() {
    //assertEquals(false, true)
  }

  @Test
  public void StringOfTest() {
    Assert.assertEquals("C#", PitchClass.Cs.toString(Sharp));
    Assert.assertEquals("Db", PitchClass.Cs.toString(Flat));
    Assert.assertEquals("-", PitchClass.Cs.toString(None));
  }


  @Test
  public void PitchClass_DiffTest() {
    Assert.assertEquals(5, PitchClass.Cs.delta(PitchClass.Fs));
    Assert.assertEquals(-5, PitchClass.Fs.delta(PitchClass.Cs));
    Assert.assertEquals(2, PitchClass.Gs.delta(PitchClass.As));
    Assert.assertEquals(-3, PitchClass.C.delta(PitchClass.A));
    Assert.assertEquals(4, PitchClass.D.delta(PitchClass.Fs));
    Assert.assertEquals(-6, PitchClass.F.delta(PitchClass.B));
  }

  private void assertPitchClassOf(String name, PitchClass expectPitchClass, String expectStringSharp, String expectStringFlat) throws Exception {
    PitchClass pitchClass = PitchClass.of(name);
    assertEquals(expectPitchClass, pitchClass);
    assertEquals(expectStringSharp, pitchClass.toString(Sharp));
    assertEquals(expectStringFlat, pitchClass.toString(Flat));
  }

}

File: engine/test/music/PitchClassTest.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <gtest/gtest.h>

#include "xjmusic/music/PitchClass.h"

using namespace XJ;

void assertPitchClassOf(
    const std::string &name,
    const PitchClass expectPitchClass,
    const std::string &expectStringSharp,
    const std::string &expectStringFlat
) {
  const PitchClass pitchClass = pitchClassOf(name);
  ASSERT_EQ(expectPitchClass, pitchClass) << name << " -> " << stringOf(pitchClass, Natural);
  ASSERT_EQ(expectStringSharp, stringOf(pitchClass, Sharp)) << name << " -> " << stringOf(pitchClass, Sharp);
  ASSERT_EQ(expectStringFlat, stringOf(pitchClass, Flat)) << name << " -> " << stringOf(pitchClass, Flat);
}

TEST(Music_PitchClass, PitchClassOf) {
  assertPitchClassOf("C", C, "C", "C");
  assertPitchClassOf("C#", Cs, "C#", "Db");
  assertPitchClassOf("Cb", B, "B", "B");
  assertPitchClassOf("D", D, "D", "D");
  assertPitchClassOf("D#", Ds, "D#", "Eb");
  assertPitchClassOf("D♭", Cs, "C#", "Db");
  assertPitchClassOf("E", E, "E", "E");
  assertPitchClassOf("E#", F, "F", "F");
  assertPitchClassOf("E♭", Ds, "D#", "Eb");
  assertPitchClassOf("F", F, "F", "F");
  assertPitchClassOf("F#", Fs, "F#", "Gb");
  assertPitchClassOf("F♭", E, "E", "E");
  assertPitchClassOf("G", G, "G", "G");
  assertPitchClassOf("G♯", Gs, "G#", "Ab");
  assertPitchClassOf("Gb", Fs, "F#", "Gb");
  assertPitchClassOf("A", A, "A", "A");
  assertPitchClassOf("A#", As, "A#", "Bb");
  assertPitchClassOf("Ab", Gs, "G#", "Ab");
  assertPitchClassOf("B", B, "B", "B");
  assertPitchClassOf("B#", C, "C", "C");
  assertPitchClassOf("E♭", Ds, "D#", "Eb");
  assertPitchClassOf("Bb", As, "A#", "Bb");
  assertPitchClassOf("z", Atonal, "X", "X");
  assertPitchClassOf("zzzz", Atonal, "X", "X");
  assertPitchClassOf("C minor", C, "C", "C");
  assertPitchClassOf("C#5", Cs, "C#", "Db");
}

TEST(Music_PitchClass, StringOf) {
  ASSERT_EQ("C#", stringOf(PitchClass::Cs, Sharp));
  ASSERT_EQ("Db", stringOf(PitchClass::Cs, Flat));
  ASSERT_EQ("C#", stringOf(PitchClass::Cs, Natural));
}